crates/sdl-encoder/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+        self.buf
+    }
+}
+
+impl Default for Schema {
+    fn default() -> Self {
+        Schema::new()
+    }
+}
+
+/// Type Definition used to define different types in SDL.
+#[derive(Debug)]
+pub struct TypeDef {
+    name: String,
+    description: Option<String>,
+    fields: Vec<Field>,
+}
+
+impl TypeDef {
+    /// Create a new instance of TypeDef with a name.
+    pub fn new(name: String, field: Field) -> Self {
+        Self {
+            name,
+            description: None,
+            fields: vec![field],
+        }
+    }
+
+    /// Set the TypeDef's description field.
+    pub fn description(&mut self, description: String) {
+        self.description = Some(description)
+    }
+
+    /// Push a Field to type def's fields vector.
+    pub fn field(&mut self, field: Field) {
+        self.fields.push(field)
+    }
+}
+
+impl Display for TypeDef {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            writeln!(f, "\"\"\"\n{}\n\"\"\"", description)?;
+        }
+
+        let mut fields = String::new();
+        for field in &self.fields {
+            fields += &format!("\n{}", field.to_string());
+        }
+
+        write!(f, "type {} {{", &self.name)?;
+        write!(f, "{}", fields)?;
+        writeln!(f, "\n}}")
+    }
+}
+
+/// A definition used when a root GraphQL type differs from default types.
+#[derive(Debug)]
+pub struct SchemaDef {
+    description: Option<String>,
+    fields: Vec<Field>,
+}
+
+impl SchemaDef {
+    /// Create a new instance of SchemaDef.
+    pub fn new(field: Field) -> Self {
+        Self {
+            description: None,
+            fields: vec![field],
+        }
+    }
+
+    /// Set the schema def's description.
+    pub fn description(&mut self, description: String) {
+        self.description = Some(description)
+    }
+
+    /// Push a Field to schema def's fields vector.
+    pub fn field(&mut self, field: Field) {
+        self.fields.push(field)
+    }
+}
+
+impl Display for SchemaDef {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            writeln!(f, "\"\"\"\n{}\n\"\"\"", description)?;
+        }
+
+        let mut fields = String::new();
+        for field in &self.fields {
+            fields += &format!("\n{}", field.to_string());
+        }
+
+        write!(f, "schema {{")?;
+        write!(f, "{}", fields)?;
+        writeln!(f, "\n}}")
+    }
+}
+
+/// Define whether a Field is a vector.
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum VecState {
+    /// The type is not a vector: e.g. `String`.
+    None,
+    /// The type is a nullable vector: e.g. `[String]`.
+    Nullable,
+    /// The type is a non nullable vector: e.g. `[String]!`.
+    NonNullable,
+}
+/// Field in a given SDL type.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Field {
+    description: Option<String>,
+    name: String,
+    type_: String,
```

Comment:
You probably don't want a type represented as a `string` here, but in a more structured way, like in [`graphql-js`'s `GraphQLType` hierarchy](https://github.com/graphql/graphql-js/blob/735975d4904c5d9cc93d0c0a02af181e8f02dfba/src/type/definition.d.ts#L36).